pocld/shared_cl_context.cc

```diff
@@ -1057,14 +1074,127 @@ int SharedCLContext::freeQueue(uint32_t queue_id) {
+                              std::vector<char> &NewSPV,
+                              const char *BuildOptions) {
+
+  // Just invoke command line tools for now.
+  constexpr int CmdOutputMaxSize = 10000;
+  char CmdOutput[CmdOutputMaxSize];
+  size_t CmdOutputSize = CmdOutputMaxSize;
+
+  char *TempDirName = strdup(
+      (std::filesystem::temp_directory_path() / "pocl-r-XXXXXX").c_str());
+
+  mkdtemp(TempDirName);
+
+  std::filesystem::path TempDir(TempDirName);
+
+  free(TempDirName);
+
+  const std::string OrigBcFileName = TempDir / "original.bc";
+
+  if (Src != nullptr && SrcSize > 0) {
+    // Compile from sources.
+    const std::string SrcFileName = TempDir / "input.cl";
+    std::ofstream SrcFile(SrcFileName);
+    SrcFile.write(Src, SrcSize);
+    SrcFile.close();
+
+    // https://www.khronos.org/blog/offline-compilation-of-opencl-kernels-into-
+    // spir-v-using-open-source-tooling
+    std::stringstream OpenCLCCmd;
+    OpenCLCCmd << CLANG
+               << " -c -target spir64 -cl-kernel-arg-info -cl-std=CL3.0 "
+               << SrcFileName.c_str() << " " << BuildOptions
+               << " -emit-llvm -o " << OrigBcFileName.c_str();
+
+    if (system(OpenCLCCmd.str().c_str()) != EXIT_SUCCESS)
+      return false;
+
+  } else if (InputSPV != nullptr) {
+    const std::string OrigSpvFileName = TempDir / "original.spv";
+
+    std::ofstream OrigSpvFile(OrigSpvFileName);
+    OrigSpvFile.write((const char *)InputSPV->data(), InputSPV->size());
+    OrigSpvFile.close();
+
+    std::stringstream SpvCmd;
+
+    SpvCmd << LLVM_SPIRV << " -r " << OrigSpvFileName.c_str() << " -o "
+           << OrigBcFileName.c_str();
+
+    if (system(SpvCmd.str().c_str()) != EXIT_SUCCESS)
+      return false;
+
+  } else {
+    assert(false && "Unimplemented.");
+  }
+
+  const std::string OffsettedBcFileName = TempDir / "offsetted.bc";
+
+  std::filesystem::path LibPoCLPath;
+  std::stringstream OptCmd;
+
+  if (!pocl_get_bool_option("POCL_BUILDING", 0))
+    LibPoCLPath /= std::filesystem::path(POCL_INSTALL_LIBDIR) / "libpocl.so";
+  else
+    LibPoCLPath /=
+        std::filesystem::path(BUILDDIR) / "lib" / "CL" / "libpocl.so";
+
+  OptCmd << LLVM_OPT << " -load-pass-plugin=" << LibPoCLPath
+         << " -passes=svm-offset -svm-offset-value=" << SVMOffset << " "
+         << OrigBcFileName << " -o " << OffsettedBcFileName;
+
+  if (system(OptCmd.str().c_str()) != EXIT_SUCCESS)
+    return false;
+
+  const std::string OutSpvFileName = TempDir / "offsetted.spv";
+
+  std::stringstream SpvCmd;
+
+  SpvCmd << LLVM_SPIRV << " " << OffsettedBcFileName.c_str() << " -o "
+         << OutSpvFileName.c_str();
+
+  if (system(SpvCmd.str().c_str()) != EXIT_SUCCESS)
+    return false;
+
+  std::ifstream OutFile(OutSpvFileName);
+  char C;
+  NewSPV.clear();
+  while (OutFile.read((char *)&C, 1)) {
+    NewSPV.push_back(C);
+  }
+
+  // std::filesystem::remove_all(TempDir);
+  return true;
+}
+
+int SharedCLContext::buildOrLinkProgram(
```

Comment:
This `buildOrLinkProgram` is quickly approaching a rather unwieldy size, I wonder if it could be cleanly split up a bit.

Reply:
I agree. Something to the "laundry list".